pkg/api/types.go

```diff
@@ -2119,34 +2119,121 @@ const (
 	StrategicMergePatchType PatchType = "application/strategic-merge-patch+json"
 )
 
-// Type and constants for component health validation.
+// Type and constants for component type.
+type ComponentType string
+
+// ComponentType constants define some known types of components.
+// Each type may have multiple instances, as long as their names are unique.
+const (
+	ComponentAPIServer         ComponentType = "apiserver"
+	ComponentControllerManager ComponentType = "controller-manager"
+	ComponentScheduler         ComponentType = "scheduler" // TODO: move mesos-specific component type & validation to contrib
+)
+
+// ComponentSpec describes the attributes that a component is created with.
+type ComponentSpec struct {
+	// Type of the component
+	Type ComponentType `json:"type"`
+	// Address to reach the component (scheme, host, port, path). Address paths determined by component type.
+	Address string `json:"address,omitempty"`
+}
+
+type ComponentPhase string
+
+// These are the valid phases of a component.
+const (
+	// ComponentPending means the component has been created, but is still starting up.
```

Comment:
I tried to make the Component API consistent with the Pod and Node APIs. I figured Pending would indicate that the Component is expected to exist, but isn't expected to be healthy/ready yet. A component still starting up may require different response/behavior/timeout from a component that has previously been healthy/ready but is no longer healthy (responding unhealthy, 500 error) or no longer responding (404, request/response/socket timeout). I meant to also distinguish between those last two, but didn't put it in the API. I figure that would be a Condition, rather than a Phase. Phase is about expectations. Conditions are about state.